zephyr/lib/regions_mm.c

```diff
@@ -1,53 +1,339 @@
 // SPDX-License-Identifier: BSD-3-Clause
 /*
- * Copyright(c) 2022 Intel Corporation. All rights reserved.
+ * Copyright(c) 2023 Intel Corporation. All rights reserved.
  *
  * Author: Jakub Dabek <jakub.dabek@intel.com>
  */
 
 #include <sof/lib/regions_mm.h>
 
-struct virtual_memory_heap
-	vm_heaps[CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT];
 
 /**
- * @brief Fill vm_heaps array with information from zephyr.
- *
- * Virtual memory regions calculated in zephyr are translated here
- * to a struct that will keep all information on current allocations
- * and virtual to physical mappings that are related to heaps.
- * System heap is not a part of this information. It only refers to
- * virtual first heaps.
- * Has to be initialized after calculations for regions is done in zephyr.
+ * @brief Initialize new heap
+ *
+ * Initializes new heap based on region that heap is to be assigned to
+ * and config.
+ * 
+ * @param cfg Configuration structure with block structure for the heap
+ * @param memory_region_attribute A virtual region identifier
+ * @param span Flag deciding if heap can do contiguous allocation.
+ *
+ * @retval ptr to a new heap if successful.
+ * @retval NULL on creation failure.
  */
-static int virtual_heaps_init(const struct device *unused)
+struct virtual_memory_heap* init_heap(struct heap_config* cfg, int memory_region_attribute, bool span)
 {
-	ARG_UNUSED(unused);
-
 	struct sys_mm_drv_region *virtual_memory_regions =
 		(struct sys_mm_drv_region *)sys_mm_drv_query_memory_regions();
+	struct virtual_memory_heap *new_heap = rzalloc(SOF_MEM_ZONE_SYS, 0,
+		SOF_MEM_CAPS_RAM, sizeof(struct virtual_memory_heap));
+	
+	/* Initialize list that will keep all active allocations*/
+	sys_slist_init(&new_heap->active_allocations);
 
+	/* Search for matching attribute so we place heap on right virtual region.
+	 * In case of core heaps regions we count theme from 0 to max avalible cores
+	 */
 	for (size_t i = 0;
 		i < CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT;
 		i++) {
-		vm_heaps[i].virtual_region = &virtual_memory_regions[i];
 
-		switch (virtual_memory_regions[i].attr)	{
-		case MEM_REG_ATTR_CORE_HEAP:
-			vm_heaps[i].memory_caps =
-				SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_HP | SOF_MEM_CAPS_CACHE;
+		if (memory_region_attribute == virtual_memory_regions[i].attr)
+		{
+			if (memory_region_attribute == MEM_REG_ATTR_CORE_HEAP)
+			{
+				new_heap->virtual_region = &virtual_memory_regions[cpu_get_id()];
+			}
+			new_heap->virtual_region = &virtual_memory_regions[i];
+			break;
+		}
+	}
+	
+	/* If no config was specified we will use default one */
+	if (cfg == NULL)
+	{
+		struct heap_config new_config = get_default_heap_config(new_heap->virtual_region);
+		cfg = &new_config;
+	}
+
+	/* Check if configuration provided can fit into virtual regions memory */
+	uint32_t total_requested_size = 0;
+	for (size_t i = 0; i < MAXIMUM_BLOCK_SIZES; i++)
+	{
+		if (cfg->block_bundles_table[i].block_size > 0)
```

Comment:
do we need this if?

Reply:
There may be unused allocator spaces. Unused spaces will be initialized to 0 and that is what this check does.